### ISSUE TITLE
Browser-based editor for Umple based on Orion

### ISSUE THREAD

**Issue by vahdat-ab** [umple/umple#707]
[Orion ](https://orionhub.org/edit/edit.html) is a browser-based IDE which provides a rich editor for web developers. It's an open source project under the Eclipse license and have been used in the next generation of Eclipse for cloud called Eclipse [Che](https://eclipse.org/che/). We could extend this editor as an advanced version of UmpleOnline.
The editor should support the current features of UmpleOnline in addition to new features such as
1. Rich text editor, which is really important because Umple is a text-based modeling language.
2. Integrated git features
3. Ability to list genarted files and see them under different folders (namespace)
4. Different themes, which is really handy for people with vision constraints
5. Editing the files in different pages
   6 . Running the server locally and get editing benefits in a local machine
   7 . Running generated code

Note: we should itemize each features (even available in mpleonline) and document them here to make sure they are feasible and also have a blueprint of the final product.
You can try the free version of Orion [here ](orionhub.org)

**Comment by vahdat-ab** [comment]
Comments from @TimLethbridge 
-  Key challenges will be to ensure that what we deliver is usable, so not thrown away at the end of the semester because it is too incomplete and too inferior to UmpleOnline.
- we have worked so incredibly hard on Umpleonline to make it simple (no modal dialogs, not too many 'in your face' options, no configuration, lots of builtin examples, command-keys to quickly jump from view to view, ability to open examples from the user manual, and so on). The lessons of Umpleonline, and where possible its look and feel must be forefront of any design.

**Comment by vahdat-ab** [comment]
To @TimLethbridge 
I suggest that you explore [orionhub ](orionhub.org) a little bit by yourself and see if you see a value on it (It's free). There might be some critical views that you may not like, so it's better not to invest on this idea.

**Comment by vtrlx** [comment]
A bit late, but I've been sorting through the viability of Orion as a platform to replace UmpleOnline. As I've pasted in my log entries, I've noted that Orion has:
- Existing support for many highlighters (Java, C++, etc...) which can be used for extra code in Umple
- Git support
- Capabilities to run projects as web applications
- A simple API for Javascript plugins

And I've also noted that the following should be possible for an Umple plugin by the end of the term:
- Provide syntax highlight for textual Umple models and extra code of other languages.
- Allow the user to compile Umple code to all the usual suspects (Java, etc)
- Allow the user to generate a visual model from their Umple code (much like how Orion does by default with Markdown)

I'm currently looking at a prototype of a plugin for a past version of Orion, to get going and will hopefully have a working highlighter soon.

**Comment by TimLethbridge** [comment]
I just tried to use OrionHub and am surprised by the lack of usability. 

After logging on the Safari browser in I am faced with a black screen with a few menu items that appear to do nothing! There is a little icon in the top right looking like a person. This ought to mean account information, but it only has one menu option 'keyboard shortcuts;' And the 'New' menu is missing in the file dialog. It might be that I created an account ages ago and have forgotten (but it should at least give me an account page that I can look at), or it might be a total fail of it to work in Safari

So I try Chrome instead, A little better, It still seems awkward. I will use it as a case study in my graduate course in usability today.

Anyway back to this issue. To proceed in an agile manner, I think we shouldn't talk about 'replacing UmpleOnline'. Instead we should talk about enabling Umple in Orion. If Umple in Orion gets good enough, only then would we consider replacing UmpleOnline. 

I agree with the three points in the above post as the first steps.

**Comment by vahdat-ab** [comment]
@VictoriaLacroix  Great news. just a comment regarding sytax highlighting of extra code. As you know we use for example "Java {", "Cpp {" for extrac codes of related codes. They migh help you much better to deal with these cases.

**Comment by vahdat-ab** [comment]
I think orionhub recently became slow because many people may want to try it. It's better to try it local because we will finally put it in our own servers.
It can e download from [here ](http://download.eclipse.org/orion/).
It woks properly for me on Chrome, Firefox, Windows edge. It's also much faster when it's local

**Comment by vtrlx** [comment]
Baby steps.

![progress](https://cloud.githubusercontent.com/assets/8976090/12896243/f47141b6-ce6e-11e5-9d93-9fb89cfd0d54.png)

Up next is better contextual highlighting, and then getting towards compiling and running with Orion.

**Comment by TimLethbridge** [comment]
Good work so far!

**Comment by Nava2** [comment]
I just realized, I meant to and forgot to point this out @VictoriaLacroix, here are some language definitions: https://github.com/umple/umple.sublime

If you have improvements, please make them and then we can use them in Sublime. It would make @aforward very happy. :wink:

**Comment by vahdat-ab** [comment]
Thanks.
The name of Plugin can be just "Umple (Model-Oriented Programming)".

**Comment by aforward** [comment]
I am now split between sublime and atom.

andrew f

On Mon, Feb 8, 2016 at 4:50 PM, Vahdat Abdelzad notifications@github.com
wrote:

> Thanks.
> The name of Plugin can be just "Umple (Model-Oriented Programming)".
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/umple/umple/issues/707#issuecomment-181571994.

## 

aforward

**Comment by vahdat-ab** [comment]
hopefully, Orion will provide a nice cloud based environment for Umple :-)

**Comment by vtrlx** [comment]
Here it is: http://github.com/VictoriaLacroix/umple.orion

**Comment by Nava2** [comment]
@VictoriaLacroix could you make me an owner of the repository? I'll move it over to the `umple` org and give you owner rights! That way it's all happening under the same organization. 

Could you try to do this yourself first, though? Under "Repository Settings" it'll be "Change Owner."

**Comment by vtrlx** [comment]
Okay.

I've transferred ownership to you, as Github didn't like when I tried giving Umple the repository.

**Comment by vahdat-ab** [comment]
Thanks. It works very well. It's gonna bring lots of value to Umple. We're looking forward to more features :-)

**Comment by Nava2** [comment]
I have now transferred it to its new home, [umple/umple.orion](https://github.com/umple/umple.orion). :+1: You should have full access, @VictoriaLacroix.

**Comment by vahdat-ab** [comment]
@VictoriaLacroix  If you go to the **plug-ins** section of Orion and then click on **get Plug-ins** at the top-right corner, you will be directed to a page in which you can find a tool called "TextMate-to-Orion grammar converter". This may be helpful (regarding a better understanding of Orion plugin architecture) because @Nava2 indicated about a grammar developed before.

**Comment by TimLethbridge** [comment]
This was started as a separate umple-orion project, but was never completed. Likely needs a fresh start.

**Comment by vahdat-ab** [comment]
yes. However, we now know more about the best approach to adopt for this issue.

**Comment by edmundluong** [comment]
I have chosen this issue as my term project.

As Vahdat mentions above, the goal is to extend Eclipse Orion, a browser-based IDE, and create an advanced version of UmpleOnline.

**Comment by vahdat-ab** [comment]
Great. I wonder if you can arrange to have an online meeting with me? There is some knowledge about this project that needs to be transformed and also we have some concerns to talk. I think this will help to start the project faster.

**Comment by edmundluong** [comment]
Absolutely, I will contact you shortly.

**Comment by TimLethbridge** [comment]
Orion is no longer a project we want to pursue
